InputRenderer: inputs for forms and grids

labkey-nicka opened the pull request:
#### Rationale
This set of changes seeks to improve our handling for custom form inputs that appear in both our forms and our editable grids. Issues addressed or recovered:

* [Issue 45486](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=45486): Biologics: Bulk Editing of Assay Run Properties
* [Issue 46488](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=46488): Typing anything into the alias field while deriving samples or media crashes the app

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/1003
* https://github.com/LabKey/biologics/pull/1694
* https://github.com/LabKey/sampleManagement/pull/1346
* https://github.com/LabKey/inventory/pull/590
* https://github.com/LabKey/labbook/pull/310

#### Changes
* Refactor `resolveInputRenderer` and enhance typings with `InputRendererFactory`.
* Update usages of `resolveRenderer` to resolve a component which can be rendered inline.
* Introduce defined types for `SelectInputChange` and `QuerySelectChange`.
* Coalesce handling of single and multi-value lookup cell change events into `onCellSelectChange`. This allows custom defined lookup components (i.e. those that are not created via `LookupCell`) to reuse the same logic for computing the new grid value upon change.
* Extract constant props for `SelectInput`s rendered to editable grid cells. This includes `customStyles` and `customThemes`.
* Update `AliasInput` and `AssayTaskInput` to extend `SelectInputProps` and extract props specific to the component.
* Update the `onChange` prop signature for `SelectInput` to provide the components props rather than a reference to the underlying `ReactSelect` instance.
* Convert `_queryColumnRenderers` to a `Record<string, any>` from an `Immutable.Map<string, any>`. Applications would use `fromJS()` which was converting renderers into `Immutable.Map` instances which are not valid React objects. Switching to `Record<string, any>` alleviates this problem by removing Immutable from the equation.

labkey-nicka committed:
Refactor resolveRenderer into InputRenderer component

labkey-nicka committed:
Rename renderers to InputRenderer

labkey-nicka committed:
Add warning

labkey-nicka committed:
SelectInput / QuerySelect: consolidate onChange handling

labkey-nicka committed:
WithFormsyProps

labkey-nicka committed:
Cell: directly supply props

labkey-nicka committed:
AncestorRenderer: lint

labkey-nicka committed:
AssayTaskInput: handle errors/loading state within SelectInput

labkey-nicka committed:
DesignerDetailPanel: refine implementation

labkey-nicka committed:
Refactor into input render factory
- this approach aligns better with our needs for fall-through rendering

labkey-nicka committed:
Vanquish isDetailInput

labkey-nicka committed:
AssayTaskInput: fix loading

labkey-nicka committed:
Grid: convert columnRenderers to a Record<string, any>
- formerly was an Immutable.Map which caused rendering errors with functional components

labkey-nicka committed:
Update docs

labkey-nicka committed:
Consolidate implementation

labkey-nicka committed:
Rename to InputRenderFactory.ts

labkey-nicka committed:
Unit tests

labkey-nicka committed:
v2.243.1-fb-input-renderer.0

cnathe commented on packages/components/src/internal/components/forms/input/AssayTaskInput.tsx:
```diff
@@ -1,40 +1,40 @@
             Filter.create('Status/Value', 'In Progress'),
         ],
         maxRows: -1,
+        schemaQuery: SchemaQuery.create('samplemanagement', 'Tasks', ViewInfo.DETAIL_NAME),
```
I assume this was intentional to update this to use the `ViewInfo. DETAIL_NAME`, just checking

labkey-nicka replied:
Yes, intentional.

cnathe commented on packages/components/src/internal/components/forms/input/AssayTaskInput.tsx:
```diff
@@ -45,82 +45,100 @@ async function loadInputOptions(assayId: number): Promise<InputOption[]> {
+            disabled={error ? true : props.disabled}
+            isLoading={loading}
+            options={taskOptions}
+            placeholder={error ? `Error: ${error}` : props.placeholder}
```
I haven't see this "disable on error and put error as placeholder text" before. Do we do this in other places for SelectInput?

labkey-nicka replied:
Yes, mainly `QuerySelect` which has been doing this since its inception.

cnathe commented on packages/components/src/internal/components/forms/input/FileInput.tsx:
```diff
@@ -26,7 +27,7 @@ import { FileColumnRenderer } from '../../../renderers/FileColumnRenderer';
 import { DisableableInput, DisableableInputProps, DisableableInputState } from './DisableableInput';
 
-interface Props extends DisableableInputProps {
+interface Props extends DisableableInputProps, WithFormsyProps {
```
this consolidation of `WithFormsyProps` is great. we should have done that earlier

cnathe commented on packages/components/src/internal/components/forms/input/SampleStatusInput.tsx:
```diff
@@ -9,118 +8,67 @@ import {
 
-// Move somewhere more central?
-// Styles to match form-control in bulk form
-export const customBulkStyles = {
```
I didn't pull down the PR to do manual testing on this. Did these customBulkStyles get rolled into the shared styles object somewhere or are they no longer needed?

labkey-nicka replied:
Thanks for pointing this out. I went back and reviewed this and ended up chatting with Marty who originally introduced the `customBulkStyles`. It turns out it was the disabled state of `SelectInput` was slightly out of sync with the disabled styles of our other form elements. I've addressed this with 2225382f55 at the `SelectInput` level by updating our default style overrides for react-select.

labkey-nicka committed:
SelectInput: improve matching of disabled styles

labkey-nicka committed:
DetailDisplay: default to not showing labels
- improve performance by not rendering labels as opposed to rendering and hiding with CSS

labkey-nicka committed:
Update test snapshots

labkey-nicka committed:
Support onSelectChange

labkey-nicka committed:
v2.243.1-fb-input-renderer.1

labkey-nicka committed:
lint

labkey-nicka committed:
Prepare release notes

labkey-nicka committed:
v2.244.0
